### PR TITLE
[IOTDB-716] add lZ4 compress method

### PR DESCRIPTION
**Pull request by SilverNarcissus**
Support lZ4 compress method

### CODE REVIEW

File: tsfile/pom.xml [hunk 1]
```diff
@@ -53,6 +53,11 @@
             <groupId>commons-io</groupId>
             <artifactId>commons-io</artifactId>
         </dependency>
+        <dependency>
+            <groupId>net.jpountz.lz4</groupId>
+            <artifactId>lz4</artifactId>
+            <version>1.3.0</version>
+        </dependency>
```
Comment (HTHou) [inline tsfile/pom.xml:60]: I think we should update the LICENSE-binary as well. :)

Comment (SilverNarcissus) [reply tsfile/pom.xml:60]: Sure~ I will fix it~

**Comment by jixuan1989** [comment]
SnappyUnCompressor.java 
```java
    @Override
    public int uncompress(byte[] byteArray, int offset, int length, byte[] output, int outOffset)
        throws IOException {
      Snappy.uncompressedLength(byteArray, offset, length);
      return Snappy.uncompress(byteArray, offset, length, output, outOffset);
    }
```
`Snappy.uncompressedLength(byteArray, offset, length);` is not used... either remove it or check whether the length of the output[] satisfies.

**Comment by jixuan1989** [comment]
besides, when something error in uncompression progress, how about throwing ioexception.

**Comment by SilverNarcissus** [comment]
> SnappyUnCompressor.java
> 
> ```java
>     @Override
>     public int uncompress(byte[] byteArray, int offset, int length, byte[] output, int outOffset)
>         throws IOException {
>       Snappy.uncompressedLength(byteArray, offset, length);
>       return Snappy.uncompress(byteArray, offset, length, output, outOffset);
>     }
> ```
> 
> `Snappy.uncompressedLength(byteArray, offset, length);` is not used... either remove it or check whether the length of the output[] satisfies.

Sure, fixed~

**Comment by SilverNarcissus** [comment]
> besides, when something error in uncompression progress, how about throwing ioexception.

Nice suggestion~ fixed~

**Comment by SilverNarcissus** [comment]
> Don't forget to resolve conflicts : )

Sure~ Fixed~

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java [hunk 2]
```diff
@@ -52,6 +52,8 @@ static ICompressor getCompressor(CompressionType name) {
         return new NoCompressor();
       case SNAPPY:
         return new SnappyCompressor();
+      case LZ4:
+        return new IOTDBLZ4Compressor();
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java:56]: To be consistent with LZ4UnCompressor and simplified, rename this to LZ4Compressor?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java [hunk 3]
```diff
@@ -184,4 +186,82 @@ public CompressionType getCodecName() {
...
+    @Override
+    public byte[] uncompress(byte[] bytes) {
+      if (bytes == null) {
+        return new byte[0];
+      }
+
+      try {
+        return decompressor.decompress(bytes, MAX_COMPRESS_RATIO * bytes.length);
+      } catch (RuntimeException e) {
+        logger.error(
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java:227]: I think it's better to throw this exception and let the upper level handle it.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java [hunk 4]
```diff
@@ -184,4 +186,82 @@ public CompressionType getCodecName() {
...
+    public int uncompress(ByteBuffer compressed, ByteBuffer uncompressed) {
+      if (compressed == null || !compressed.hasRemaining()) {
+        return 0;
+      }
+
+      try {
+        decompressor.decompress(compressed, uncompressed);
+        return compressed.limit();
+      } catch (RuntimeException e) {
+        logger.error(
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java:256]: throw IOException

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java (hunk 2 above)
Comment (SilverNarcissus) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java:56]: The package I import is called LZ4Compressor, so I can't use this name

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java (hunk 3 above)
Comment (SilverNarcissus) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java:227]: Sure~

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java (hunk 4 above)
Comment (SilverNarcissus) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java:256]: Sure~

**Pull request merged**